swarm/storage/ldbstore.go

```diff
@@ -255,98 +235,49 @@ func decodeOldData(data []byte, chunk *Chunk) {
 	chunk.Size = int64(binary.BigEndian.Uint64(data[0:8]))
 }
 
-func gcListPartition(list []*gcItem, left int, right int, pivotIndex int) int {
-	pivotValue := list[pivotIndex].value
-	dd := list[pivotIndex]
-	list[pivotIndex] = list[right]
-	list[right] = dd
-	storeIndex := left
-	for i := left; i < right; i++ {
-		if list[i].value < pivotValue {
-			dd = list[storeIndex]
-			list[storeIndex] = list[i]
-			list[i] = dd
-			storeIndex++
-		}
-	}
-	dd = list[storeIndex]
-	list[storeIndex] = list[right]
-	list[right] = dd
-	return storeIndex
-}
-
-func gcListSelect(list []*gcItem, left int, right int, n int) int {
-	if left == right {
-		return left
-	}
-	pivotIndex := (left + right) / 2
-	pivotIndex = gcListPartition(list, left, right, pivotIndex)
-	if n == pivotIndex {
-		return n
-	} else {
-		if n < pivotIndex {
-			return gcListSelect(list, left, pivotIndex-1, n)
-		} else {
-			return gcListSelect(list, pivotIndex+1, right, n)
-		}
-	}
-}
-
 func (s *LDBStore) collectGarbage(ratio float32) {
 	it := s.db.NewIterator()
-	it.Seek(s.gcPos)
-	if it.Valid() {
-		s.gcPos = it.Key()
-	} else {
-		s.gcPos = nil
-	}
-	gcnt := 0
+	defer it.Release()
 
-	for (gcnt < gcArraySize) && (uint64(gcnt) < s.entryCnt) {
+	garbage := []*gcItem{}
+	gcnt := 0
 
-		if (s.gcPos == nil) || (s.gcPos[0] != kpIndex) {
-			it.Seek(s.gcStartPos)
-			if it.Valid() {
-				s.gcPos = it.Key()
```

Comment:
FYI this is one of the bugs in the garbage collector - `it.Key()` gets modified on each `it.Next()` call, and so is `s.gcPos` - therefore the `idxKey` is a single entry in the `gcArray` and we would remove at most 1 entry per call to `collectGarbage()`

Reply:
genious